packages/vite/src/node/plugins/css.ts

```diff
@@ -924,9 +931,49 @@ async function compileCSS(
   }
 
   if (isModule) {
+    const FileSystemLoader = // @ts-expect-error TODO: needs types
+      (await import('postcss-modules/build/FileSystemLoader')).default
     postcssPlugins.unshift(
       (await import('postcss-modules')).default({
         ...modulesOptions,
+        Loader: class extends FileSystemLoader {
+          async fetch(
+            _newPath: string,
+            relativeTo: string,
+            _trace: string,
+          ): Promise<any> {
```

Comment:
base logic is copied from https://github.com/madyankin/postcss-modules/blob/325f0b33f1b746eae7aa827504a5efd0949022ef/src/FileSystemLoader.js#L68

I added preprocessing between L947 and L970
https://github.com/vitejs/vite/pull/12076/files#diff-2cfbd4f4d8c32727cd8e1a561cffbde0b384a3ce0789340440e144f9d64c10f6R947-R970